crates/manifest/src/schema/v2.rs

```diff
@@ -82,12 +87,123 @@ pub struct Trigger {
 }
 
 /// One or many `ComponentSpec`(s)
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
 #[serde(transparent)]
-pub struct OneOrManyComponentSpecs(#[serde(with = "one_or_many")] pub Vec<ComponentSpec>);
+pub struct OneOrManyComponentSpecs(
+    #[serde(with = "one_or_many")]
+    #[schemars(schema_with = "one_or_many_schema::<ComponentSpec>")]
+    pub Vec<ComponentSpec>,
+);
+
+fn one_or_many_schema<T: schemars::JsonSchema>(
+    gen: &mut schemars::gen::SchemaGenerator,
+) -> schemars::schema::Schema {
+    schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+        subschemas: Some(Box::new(schemars::schema::SubschemaValidation {
+            one_of: Some(vec![
+                gen.subschema_for::<T>(),
+                gen.subschema_for::<Vec<T>>(),
+            ]),
+            ..Default::default()
+        })),
+        ..Default::default()
+    })
+}
+
+fn toml_table_schema(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+    schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+        instance_type: Some(schemars::schema::SingleOrVec::Single(Box::new(
+            schemars::schema::InstanceType::Object,
+        ))),
+        ..Default::default()
+    })
+}
+
+fn map_of_toml_tables_schema(
+    _gen: &mut schemars::gen::SchemaGenerator,
+) -> schemars::schema::Schema {
+    schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+        instance_type: Some(schemars::schema::SingleOrVec::Single(Box::new(
+            schemars::schema::InstanceType::Object,
+        ))),
+        ..Default::default()
+    })
+}
+
+/// TODO: there may be a way to do this without dead code using `schemars::extend` or
+/// something like that?
+#[allow(dead_code)]
+mod schema {
+    use super::{toml_table_schema, ComponentSpec, Map, OneOrManyComponentSpecs};
+    use schemars::JsonSchema;
+
+    #[derive(JsonSchema)]
+    pub struct TriggerSchema {
+        /// HTTP triggers
+        #[schemars(default)]
+        http: Vec<HttpTriggerSchema>,
+        /// Redis triggers
+        #[schemars(default)]
+        redis: Vec<RedisTriggerSchema>,
+    }
+
+    #[derive(JsonSchema)]
+    #[schemars(deny_unknown_fields)]
+    struct HttpTriggerSchema {
```

Comment:
This isn't the first time we've had to duplicate these for crate organization reasons; I wonder if it we should throw logic to the wind and move the canonical definitions of these "special" trigger configs into this crate... :grimacing:

Reply:
My puritanical soul is scandalised to hear such a thing!  But yeah the duplication is a bit regrettable.  I propose to declare it out of scope for this PR, but we should come back to it.